CanonicalABI.md

```diff
@@ -0,0 +1,1166 @@
+the reference count. They are implemented as follows:
+
+```
+def resource_clone(src, $resource, val):
+  handle = TABLE.get($resource, src, val)
+  return TABLE.insert($resource, handle, src)
+
+def resource_drop(src, $resource, val):
+  return TABLE.remove($resource, handle, src)
+```
+
+Notably `clone` and `drop` will both trap if the index is invalid and the
+instance doesn't actually have access to that resource. The `clone` function is
+used to create a new index pointing to the same resource in the module's index
+space. The `drop` function indicates that the module no longer needs access to
+that index. If the reference count reaches 0 at that time then the destructor
+for the resource is run (perhaps invoking another module).
+
+If a module defined `$resource` then it can also import these functions:
+
+```wasm
+(module
+  (import "canonical_abi" "resource_new_$resource" (func (param i32) (result i32)))
+  (import "canonical_abi" "resource_get_$resource" (func (param i32) (result i32)))
+)
+```
+
+which are implemented as:
+
+```
+def resource_new(src, $resource, val):
+  handle = HANDLE(val=val, refcnt=0, src=src) # refcnt bumped to 1 in insert next
+  return TABLE.insert($resource, handle, src)
+
+def resource_get(src, $resource, val):
+  return TABLE.get($resource, handle, src).val
+```
+
+These functions are used to create a handle wrapper around a resource
+(identified as a unique i32) and then to access the private i32 value. These
+functions are only available to modules/instances that defined the resource
+which means no other instances can have access to the private internals. Note
+that `get` will trap if the index passed in is invalid.
+
+Note that hosts may have their own reference count of resources owned by wasm as
+well. For example if a wasm resource makes its way all the way to a host then
+the host may hold a reference count that, when dropped, may trigger a wasm
+destructor. Similarly if a wasm resource is owned by the host and all other wasm
+modules with access drop the resource then the resource isn't dropped in the
+original module until the host is done with it.
+
+## Buffer Intrinsics
+
+Buffers defined in foreign modules are identified by an `i32` index. To actually
+operate on that index the following intrinsics can be used:
+
+```wasm
+(module
+  (import "canonical_abi" "push_buffer_len" (func (param i32) (result i32)))
+  (import "canonical_abi" "push_buffer_push" (func (param i32 i32 i32) (result i32)))
+  (import "canonical_abi" "pull_buffer_len" (func (param i32) (result i32)))
+  (import "canonical_abi" "pull_buffer_pull" (func (param i32 i32 i32) (result i32)))
+)
+```
+
+The implementation of these intrinsics is supplied by the table/buffer
+management glue.
+
+The `canonical_abi::push_buffer_len` function takes the integer descriptor for
+the buffer and returns the integer length of the buffer. This is the maximal
+number of items which can be pushed into the buffer.
```

Comment:
Two related questions:
1. Does the buffer length decrease over time after repeated pushes/pulls to it (reflecting progress filling the buffer)?
2. Is it bad that a pull/push-buffer of `char` is effectively UTF-32-encoded?

For 2, I think it is a problem: folks wanting something more compact would be forced to use `(push-buffer u8)`, and now we have lots of potential for subtle encoding incompatibilities.  But if we use UTF-8 we have a problem: UTF-8 pull and push buffers wouldn't have an *a priori* number-of-units length.  (Esp. once you start considering custom adapter functions with arbitrary encodings on either side.)  I wonder if a solution is to change the push/pull len to be an "estimated" length remaining, which (answering question (1)) would decrease over time.  The idea being: the recipient of the buffer can use the estimate to push/pull about the right amount, and repeatedly pull/push until the estimate returned is 0.  One kindof neat thing is that, in the future with custom adapter functions, this could allow a custom adapter to use a single call and single push buffer to, say, fill a list-of-chunks.  Also, `char` might not be the only element type that wants a variable-length encoding with custom adapter functions.

WDYT?

Reply:
For (1) yeah the buffer length is updated on each call to push/pull so subsequent calls only operate on what the remaining space in the buffer is. This is sorta in `def push` and `def pull` where they update `buffer.len` incrementally as items are moved back and forth.

For (2) you're right that `(push-buffer char)` is utf-32 right now. About the a-priori things, though, this feels a bit odd. You'd still want to construct a `char` buffer with ptr+len where len is the utf8-length, so on construction you don't know the number of `char` elements. This means that the adapter glue would have to implement utf-8 encoding/decoding to figure out the real number of `char` elements, but the `len` would be wrong for multibyte characters (always) unless the glue calculated that?

I do think it makes sense for the len to be "estimated" though in the long run. On construction though I think the `len` is required to always be exact because it's pointing in memory though?

Reply:
Right, so I was thinking that, on the canonical-lifting side, `len` does have to be the exact byte length.  Then, on the canonical lowering side, the estimate would be determined by the element type: for all types other than `char`, it would be `len/size(elem)`; for `char` it would be simply `len` (and thus wrong in the case of multi-byte, but that's ok, b/c it's just an estimate).

Reply:
Oh currently on the lifting side `len` is an element length, not a byte length. If we say that `len` on the consuming side is just a hint, does that actually change much here? Perhaps renaming `len` to `size_hint` or something like that?

Reply:
Right, so I thinking different units for producer/consumer: producer provides number of bytes, consumer receives estimated number of elements (with the estimate being derived from the number of bytes)?

Reply:
I originally avoided byte length to avoid dealing with the question "what if you give me 3 bytes for a struct that's 4 bytes large?" The obvious answer there is truncation but it seems like it would be best to not have to answer that at  all.

But I'm still not sure what is actually changing here? Bytes/elements would change the one division for calculation of the length, but otherwise are you thinking it's just documentation that's changing? (or maybe function names)?

Do you think there needs to be wording to say that `push-buffer char` is a utf-8 encoded buffer? (I'm not sure if it's really worth going to too much length here if that's not expected to be super common, which I'm not sure if it is?

Reply:
Ah, so are you thinking that, on the lifting size the given `i32` len could be "number of elements", except for `char`, where it would be "number of bytes"?  I suppose that could work too.

Reply:
Oh I'm not thinking anything per-se but that could possibly work, yeah. I'm not personally convinced of the need to use utf-8 here so I'm trying to better understand how you (who I think wants to use utf-8?) would like things to work. In the end though I'm fine with whatever!

Reply:
Actually, I think I just talked myself out of it: if we wanted a UTF-8 buffer, by symmetry to `string`, we should do so by defining a buffer+string specialization (say, `*-string-buffer`).  Until we do that, by symmetry to `list char`, `*-buffer char` should be UTF-32.  So "n/m".